CritBit/src/org/zoodb/index/critbit/CritBit64COW.java

```diff
@@ -94,113 +101,120 @@ public V put(long key, V val) {
+                        } else {
+                            parent.hiPost = subPrefix;
+                            parent.hi = newSub;
+                        }
+                        newInfo.size++;
+                        return null;
                     }
-                    n.hi = n2;
-                    n.hiPost = extractPrefix(key, n2.posDiff-1);
-                    n.hiVal = null;
-                    size++;
-                    return null;
                 }
-            } else {
-                if (n.lo != null) {
-                    prefix = n.loPost;
-                    parent = n;
-                    n = n.lo;
-                    isCurrentChildLo = true;
+
+                //prefix matches, so now we check sub-nodes and postfixes
+                if (BitTools.getBit(key, n.posDiff)) {
+                    if (n.hi != null) {
+                        prefix = n.hiPost;
+                        parent = n;
+                        n = n.hi;
+                        isCurrentChildLo = false;
+                    } else {
+                        Node<V> n2 = createNode(key, val, n.hiPost, n.hiVal);
+                        if (n2 == null) {
+                            V prev = n.hiVal;
+                            n.hiVal = val;
+                            return prev;
+                        }
+                        n.hi = n2;
+                        n.hiPost = extractPrefix(key, n2.posDiff - 1);
+                        n.hiVal = null;
+                        newInfo.size++;
+                        return null;
+                    }
                 } else {
-                    Node<V> n2 = createNode(key, val, n.loPost, n.loVal);
-                    if (n2 == null) {
-                        V prev = n.loVal;
-                        n.loVal = val;
-                        return prev;
+                    if (n.lo != null) {
+                        prefix = n.loPost;
+                        parent = n;
+                        n = n.lo;
+                        isCurrentChildLo = true;
+                    } else {
+                        Node<V> n2 = createNode(key, val, n.loPost, n.loVal);
+                        if (n2 == null) {
+                            V prev = n.loVal;
+                            n.loVal = val;
+                            return prev;
+                        }
+                        n.lo = n2;
+                        n.loPost = extractPrefix(key, n2.posDiff - 1);
+                        n.loVal = null;
+                        newInfo.size++;
+                        return null;
                     }
-                    n.lo = n2;
-                    n.loPost = extractPrefix(key, n2.posDiff-1);
-                    n.loVal = null;
-                    size++;
-                    return null;
                 }
+                parentPosDiff = n.posDiff;
             }
-            parentPosDiff = n.posDiff;
+        } finally {
+            this.info = newInfo;
```

Comment:
Hmm, shouldn't that be a compareAndSwap here? It looks like you are simply overwriting the previous value without checking whether another thread updated this in the meantime?